File: p.107/Golang/solution.go
package main

import (
  "bufio"
  "fmt"
  "os"
  "strconv"
  "strings"
  "sort"
)

type Edge struct {
  Weight int
  U int
  V int
}

type UnionFind struct {
  Parent []int
  Rank []int
}

func NewUnionFind(size int) *UnionFind {
  parent := make([]int, size)
  rank := make([]int, size)
  for i := range parent {
    parent[i] = i
  }
  return &UnionFind{Parent: parent, Rank: rank}
}

func (uf *UnionFind) Find(x int) int {
  if uf.Parent[x] != x {
    uf.Parent[x] = uf.Find(uf.Parent[x])
  }
  return uf.Parent[x]
}

func (uf *UnionFind) Union(x, y int) {
  rootX := uf.Find(x)
  rootY := uf.Find(y)

  if rootX != rootY {
    if uf.Rank[rootX] < uf.Rank[rootY] {
      uf.Parent[rootX] = rootY
    } else if uf.Rank[rootX] > uf.Rank[rootY] {
      uf.Parent[rootY] = rootX
    } else {
      uf.Parent[rootY] = rootX
      uf.Rank[rootX]++
    }
  }
}


type Graph struct {
  Vertices int
  Edges []Edge
}

func (g *Graph) AddEdge(u, v, weight int) {
  g.Edges = append(g.Edges, Edge{Weight: weight, U: u, V: v})
}

func (g *Graph) KruskalMST() int {
  sort.Slice(g.Edges, func (i, j int) bool {
    return g.Edges[i].Weight < g.Edges[j].Weight
  })


  uf := NewUnionFind(g.Vertices)
  mstWeight := 0

  for _, edge := range g.Edges {
    if uf.Find(edge.U) != uf.Find(edge.V) {
      uf.Union(edge.U, edge.V)
      mstWeight += edge.Weight
    }
  }

  return mstWeight
}

func (g *Graph) TotalWeight() int {
  total := 0
  for _, edge := range g.Edges {
    total += edge.Weight
  }
  return total
}

func ParseNetworkMatrix(filename string) (*Graph, error) {
  file, err := os.Open(filename)
  if err != nil {
    return nil, err
  }
  defer file.Close()

  scanner := bufio.NewScanner(file)
  var matrix [][]string

  for scanner.Scan() {
    line := strings.Split(scanner.Text(), ",")
    matrix = append(matrix, line)
  }

  if err := scanner.Err(); err != nil {
    return nil, err
  }

  vertices := len(matrix)
  graph := &Graph{Vertices: vertices}

  for i := 0; i < vertices; i++ {
    for j := i + 1; j < vertices; j++ {
      if matrix[i][j] != "-" {
        weight, err := strconv.Atoi(matrix[i][j])
        if err != nil {
          return nil, fmt.Errorf("Invalid weight at (%d, %d): %v", i, j, err)
        }
        graph.AddEdge(i, j, weight)
      }
    }
  }

  return graph, nil
}

func main() {
  filename := "0107_network.txt"

  graph, err := ParseNetworkMatrix(filename)
  if err != nil {
    fmt.Println("Error: ", err)
    return
  }

  originalWeight := graph.TotalWeight()
  mstWeight := graph.KruskalMST()
  savings := originalWeight - mstWeight

  fmt.Printf("Savings: %d\n", savings)
}

File: p.107/Python/solution.py
class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.edges = []

    def add_edge(self, u, v, weight):
        if weight != -1:
            self.edges.append((weight, u, v))

    def find(self, parent, i):
        if parent[i] != i:
            parent[i] = self.find(parent, parent[i])
        return parent[i]
    
    def union(self, parent, rank, x, y):
        root_x = self.find(parent, x)
        root_y = self.find(parent, y)
        if root_x != root_y:
            if rank[root_x] < rank[root_y]:
                parent[root_x] = root_y
            elif rank[root_x] > rank[root_y]:
                parent[root_y] = root_x
            else:
                parent[root_y] = root_x
                rank[root_x] += 1

    def kruskal_mst(self):
        self.edges.sort()
        parent = [i for i in range(self.V)]
        rank = [0] * self.V

        mst_weight = 0
        mst_edges = []

        for weight, u, v in self.edges:
            if self.find(parent, u) != self.find(parent, v):
                self.union(parent, rank, u, v)
                mst_weight += weight
                mst_edges.append((u, v, weight))
        return mst_weight

def parse_network_matrix(filename):
    with open(filename, 'r') as file:
        lines = file.readlines()
        matrix = [line.strip().split(',') for line in lines]
        
    n = len(matrix)
    g = Graph(n)

    for i in range(n):
        for j in range(i + 1, n):
            weight = int(matrix[i][j] if matrix[i][j] != '-' else -1)
            g.add_edge(i, j, weight)
    return g

def calculate_max_savings(filename):
    g = parse_network_matrix(filename)
    original_weight = sum(edge[0] for edge in g.edges)
    mst_weight = g.kruskal_mst()
    savings = original_weight - mst_weight
    print(savings)


if __name__ == "__main__":
    calculate_max_savings("0107_network.txt")
